standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java

```diff
@@ -678,316 +215,38 @@ private void setCleanerStart(Connection dbConn, CompactionInfo info, Long timest
-    PreparedStatement pStmt = null;
-    ResultSet rs = null;
-    try {
-      /*
-       * compaction may remove data from aborted txns above tc_writeid bit it only guarantees to
-       * remove it up to (inclusive) tc_writeid, so it's critical to not remove metadata about
-       * aborted TXN_COMPONENTS above tc_writeid (and consequently about aborted txns).
-       * See {@link ql.txn.compactor.Cleaner.removeFiles()}
-       */
-      String s = "DELETE FROM \"TXN_COMPONENTS\" WHERE \"TC_TXNID\" IN ( "
-              + "SELECT \"TXN_ID\" FROM \"TXNS\" WHERE \"TXN_STATE\" = " + TxnStatus.ABORTED + ") "
-              + "AND \"TC_DATABASE\" = ? AND \"TC_TABLE\" = ? "
-              + "AND \"TC_PARTITION\" " + TxnUtils.nvl(info.partName);
-
-      List<String> queries = new ArrayList<>();
-      Iterator<Long> writeIdsIter = null;
-      List<Integer> counts = null;
-
-      if (info.writeIds != null && !info.writeIds.isEmpty()) {
-        StringBuilder prefix = new StringBuilder(s).append(" AND ");
-        List<String> questions = Collections.nCopies(info.writeIds.size(), "?");
-
-        counts = TxnUtils.buildQueryWithINClauseStrings(conf, queries, prefix,
-                new StringBuilder(), questions, "\"TC_WRITEID\"", false, false);
-        writeIdsIter = info.writeIds.iterator();
-      } else if (!info.hasUncompactedAborts) {
-        if (info.highestWriteId != 0) {
-          s += " AND \"TC_WRITEID\" <= ?";
-        }
-        queries.add(s);
-      }
-
-      for (int i = 0; i < queries.size(); i++) {
-        String query = queries.get(i);
-        int writeIdCount = (counts != null) ? counts.get(i) : 0;
-
-        LOG.debug("Going to execute update <{}>", query);
-        pStmt = dbConn.prepareStatement(query);
-        int paramCount = 1;
-
-        pStmt.setString(paramCount++, info.dbname);
-        pStmt.setString(paramCount++, info.tableName);
-        if (info.partName != null) {
-          pStmt.setString(paramCount++, info.partName);
-        }
-        if (info.highestWriteId != 0 && writeIdCount == 0) {
-          pStmt.setLong(paramCount, info.highestWriteId);
-        }
-        for (int j = 0; j < writeIdCount; j++) {
-          if (writeIdsIter.hasNext()) {
-            pStmt.setLong(paramCount + j, writeIdsIter.next());
-          }
-        }
-        int rc = pStmt.executeUpdate();
-        LOG.debug("Removed {} records from txn_components", rc);
-      }
-    } catch (SQLException e) {
-      LOG.error("Unable to delete from txn components due to {}", e.getMessage());
-      LOG.debug("Going to rollback");
-      rollbackDBConn(dbConn);
-      checkRetryable(e, "markCleanedForAborts(" + info + ")");
-      throw new MetaException(DB_FAILED_TO_CONNECT + e.getMessage());
-    } finally {
-      close(rs);
-      closeStmt(pStmt);
-    }
+    new MarkCleanedFunction(info, dbProduct, conf).execute(jdbcTemplate);
```

Comment:
can we do 
jdbcTemplate.execute(new MarkCleanedFunction(info, dbProduct, conf)) -> 
function.execute(jdbcTemplate)

Reply:
1st: While it was a great idea for QueryHandlers and ParameterizedCommands, for the TransactionalFunction it would look like: 
```
  public <R> R execute(TransactionalFunction<R> function) throws SQLException, MetaException {
    function.execute(this);
  }

```
In this case it doesn't have much added value.

2st: fixed.